fv3core/decorators.py

```diff
@@ -97,518 +63,221 @@ def get_namespace(arg_specs, state):
-                if field_info[field_name]
-                and field_info[field_name].access != AccessKind.READ_ONLY
-            ]
-        return self._write_fields
-
-
-class FV3StencilObject(StencilWrapper):
-    """GT4Py stencil object used for fv3core."""
-
-    def __init__(self, func: Callable[..., None], **kwargs):
-        super().__init__(func, delay_compile=True)
-
-        self.times_called: int = 0
-        """Number of times this stencil has been called."""
-
-        self.timers = types.SimpleNamespace(call_run=0.0, run=0.0)
-        """Accumulated time spent in this stencil.
-
-        call_run includes stencil call overhead, while run omits it."""
-
-        self._passed_externals: Dict[str, Any] = kwargs.pop("externals", {})
-        """Externals passed in the decorator (others are added later)."""
-
-        self.backend_kwargs: Dict[str, Any] = kwargs
-        """Remainder of the arguments assumed to be compiler backend options."""
-
-        self._data_cache: StencilDataCache = StencilDataCache("data_cache.p")
-        """Data cache to store axis offsets and passed externals."""
+    def field_names(self) -> Tuple[str]:
+        """names of stencil field call arguments"""
+        return cast(Tuple[str], tuple(self.stencil_object.field_info.keys()))
 
     @property
-    def built(self) -> bool:
-        """Indicates whether the stencil is loaded."""
-        return self.stencil_object is not None
+    def parameter_names(self) -> Tuple[str]:
+        """names of stencil parameter call arguments"""
+        return cast(Tuple[str], tuple(self.stencil_object.parameter_info.keys()))
+
+
+def compute_field_origins(
+    stencil_object: gt4py.StencilObject,
+    origin: Union[Index3D, Mapping[str, Tuple[int, ...]]],
+) -> Dict[str, Tuple[int, ...]]:
+    """Computes the origin for each field in the stencil call."""
+    field_origins = stencil_object._make_origin_dict(
+        origin
+    )  # optional: if not done before call
+    all_origin = field_origins.get("_all_", None)
+
+    for name, field_info in stencil_object.field_info.items():
+        if field_info is not None:
+            field_origin = field_origins.get(name, None)
+            if field_origin is not None:
+                field_origin_ndim = len(field_origin)
+                if field_origin_ndim != field_info.ndim:
+                    assert field_origin_ndim == field_info.domain_ndim, (
+                        f"Invalid origin specification ({field_origin})"
+                        f" for '{name}' field."
+                    )
+                    field_origins[name] = (
+                        *field_origin,
+                        *((0,) * len(field_info.data_dims)),
+                    )
+            elif all_origin is not None:
+                field_origins[name] = (
+                    *gt4py.utils.filter_mask(all_origin, field_info.domain_mask),
+                    *((0,) * len(field_info.data_dims)),
+                )
```

Comment:
would you mind describing in the PR description what this change is about, what motivates it and what it fixes etc? comparing the diff from jeremy's PR, it is still not clear to me

Reply:
For now I tried removing this change again. We will need it at some point soon when we introduce higher dimensional fields, but for now I think it's ok...